src/cpp/session/modules/rmarkdown/SessionRmdNotebook.cpp

```diff
@@ -0,0 +1,670 @@
+{
+   std::vector<std::string> chunkIds;
+   extractChunkIds(chunkOutputs, &chunkIds);
+
+   // find all the chunks and play them back to the client
+   BOOST_FOREACH(const std::string& chunkId, chunkIds)
+   {
+      enqueueChunkOutput(docPath, docId, chunkId);
+   }
+
+   json::Object result;
+   result["path"] = docPath;
+   result["request_id"] = requestId;
+   ClientEvent event(client_events::kChunkOutputFinished, result);
+   module_context::enqueClientEvent(event);
+}
+
+// called by the client to inject output into a recently opened document 
+Error refreshChunkOutput(const json::JsonRpcRequest& request,
+                         json::JsonRpcResponse*)
+{
+   // extract path to doc to be refreshed
+   std::string docPath, docId, requestId;
+   Error error = json::readParams(request.params, &docPath, &docId, &requestId);
+   if (error)
+      return error;
+
+   json::Object result;
+   json::Value chunkDefs; 
+   error = getChunkDefs(docPath, docId, &chunkDefs);
+
+   // schedule the work to play back the chunks (we don't do it synchronously
+   // so the RPC can return immediately)
+   if (!error && chunkDefs.type() == json::ArrayType) 
+   {
+      module_context::scheduleDelayedWork(boost::posix_time::milliseconds(10), 
+            boost::bind(replayChunkOutputs, docPath, docId, requestId, 
+                        chunkDefs.get_array()));
+   }
+
+   return Success();
+}
+
+bool copyCacheItem(const FilePath& from,
+                   const FilePath& to,
+                   const FilePath& path)
+{
+
+   std::string relativePath = path.relativePath(from);
+   FilePath target = to.complete(relativePath);
+
+   Error error = path.isDirectory() ?
+                     target.ensureDirectory() :
+                     path.copy(target);
+   if (error)
+      LOG_ERROR(error);
+
+   return true;
+}
+
+Error copyCache(const FilePath& from, const FilePath& to)
+{
+   Error error = to.ensureDirectory();
+   if (error)
+      return error;
+
+   return from.childrenRecursive(
+             boost::bind(copyCacheItem, from, to, _2));
+}
+
+void onDocRemoved(const std::string& docId)
+{
+   // check to see if this document was an unsaved notebook, and clean up its
+   // cache folder if so
+   FilePath cacheFolder = chunkCacheFolder("", docId);
+   Error error = cacheFolder.removeIfExists();
+   if (error)
+      LOG_ERROR(error);
+}
+
+void onDocRenamed(const std::string& oldPath, 
+                  boost::shared_ptr<source_database::SourceDocument> pDoc)
+{
+   // compute cache folders and ignore if we can't safely adjust them
+   FilePath oldCacheDir = chunkCacheFolder(oldPath, pDoc->id());
+   FilePath newCacheDir = chunkCacheFolder(pDoc->path(), pDoc->id());
+   if (!oldCacheDir.exists() || newCacheDir.exists())
+      return;
+
+   // if the doc was previously unsaved, we can just move the whole folder 
+   // to its newly saved location
+   if (oldPath.empty())
+   {
+      oldCacheDir.move(newCacheDir);
```

Comment:
This might fail if it's cross-volume so you might need a fallback that copies and removes the original.